language/diem-framework/modules/CRSN.move

```diff
@@ -0,0 +1,218 @@
+/// A module implementing conflict-resistant sequence numbers (CRSNs).
+/// The specification, and formal description of the acceptance and rejection
+/// criteria, force expiration and window shifting of CRSNs are described in DIP-168.
+
+module DiemFramework::CRSN {
+    use DiemFramework::DiemVersion;
+    use DiemFramework::DiemConfig;
+    use DiemFramework::DiemTimestamp;
+    use Std::BitVector::{Self, BitVector};
+    use Std::Signer;
+    use Std::Errors;
+
+    friend DiemFramework::DiemAccount;
+
+    /// A CRSN  represents a finite slice or window of an "infinite" bitvector
+    /// starting at zero with a size `k` defined dynamically at the time of
+    /// publication of CRSN resource. The `min_nonce` defines the left-hand
+    /// side of the slice, and the slice's state is held in `slots` and is of size `k`.
+    /// Diagrammatically:
+    /// ```
+    /// 1111...000000100001000000...0100001000000...0000...
+    ///        ^             ...                ^
+    ///        |____..._____slots______...______|
+    ///     min_nonce                       min_nonce + k - 1
+    /// ```
+    struct CRSN has key {
+        min_nonce: u64,
+        size: u64,
+        slots: BitVector,
+    }
+
+    /// No CRSN resource exists
+    const ENO_CRSN: u64 = 0;
+    /// A CRSN resource wasn't expected, but one was found
+    const EHAS_CRSN: u64 = 1;
+    /// The size given to the CRSN at the time of publishing was zero, which is not supported
+    const EZERO_SIZE_CRSN: u64 = 2;
+    /// The size given to the CRSN at the time of publishing was larger than the largest allowed CRSN size
+    const ECRSN_SIZE_TOO_LARGE: u64 = 3;
+    /// CRSNs are not supported until after Diem version 4
+    const ECRSN_NOT_SUPPORTED_AT_VERSION: u64 = 4;
+    /// the amount to shift the CRSN window was zero
+    const EINVALID_SHIFT: u64 = 5;
+
+    const MAX_CRSN_SIZE: u64 = 256;
+    const SUPPORTED_AFTER_VERSION: u64 = 3;
+
+
+    /// Publish a DSN under `account`. Cannot already have a DSN published.
+    public(friend) fun publish(account: &signer, min_nonce: u64, size: u64) {
+        assert(!has_crsn(Signer::address_of(account)), Errors::invalid_state(EHAS_CRSN));
+        assert(DiemVersion::get() > SUPPORTED_AFTER_VERSION, Errors::invalid_state(ECRSN_NOT_SUPPORTED_AT_VERSION));
```

Comment:
Do we need to gate it by DiemVersion? Cuz the moment we publish this CSRN module we can already publish this resource.

Reply:
My thinking was that by preventing this resource from being published until the specified version that it gives us a way to control the release better, but if you don't think it buys much, I'm happy to remove it and simplify things